### PR TITLE
initial commit

### LINKED ISSUE #602: Ajouter un lien “Voir sur transport.data.gouv.fr” pour les jeux de données qui sont référencés sur le PAN
Cela permettra aux utilisateurs de data.gouv.fr de découvrir la plateforme thématique ainsi que d’accéder à des outils adaptés pour les données transport (visualisation, rapports de validation...).

### PR DESCRIPTION
**Pull request by quaxsze**
fix https://github.com/etalab/data.gouv.fr/issues/602

### CODE REVIEW

File: udata_transport/templates/dataset-transport.html [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{%- if transport_url %}
+<div class="bg-alt-blue-cumulus fr-p-3v fr-mb-1w">
+    <span class="fr-grid-row">
+      <span class="fr-fi-external-link-line fr-mr-1w" aria-hidden="true"></span>
+      Ce jeu de données est aussi référencé sur le portail transport.data.gouv.fr <a href={{transport_url}}>Voir.</a>
```
Comment (maudetes) [inline udata_transport/templates/dataset-transport.html]: How about? (@abulte, let me know if you have a better wording)
```suggestion
      Ce jeu de données est aussi référencé sur le portail transport.data.gouv.fr.&nbsp;<a href={{transport_url}}>Voir ce jeu de données sur le portail transport.</a>
```

File: udata_transport/tasks.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from udata.commands import success, error
+from udata.models import Dataset
+
+
+def process_dataset(dataset):
+    target_dataset = Dataset.objects(id=dataset['datagouv_id']).first()
+    if not target_dataset:
+        error(f"Dataset {dataset['id']} not found")
+        return
+    target_dataset.extras['transport_url'] = dataset['page_url']
```
Comment (maudetes) [inline udata_transport/tasks.py]: Do we unset this extras for all datasets?

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          keys:
+          - py3-cache-v1-{{ checksum "python.deps" }}
+          - py3-cache-v1-{{ .Branch }}
+          - py3-cache-v1-{{ .Environment.BASE_BRANCH }}
+      - run:
+          name: Install python dependencies
+          command: |
+            virtualenv venv
+            source venv/bin/activate
+            pip install -e . || pip install -e .
```
Comment (maudetes) [inline .circleci/config.yml]: Do we still need this weird logic? Can't we install requirement and then editable?

File: requirements/install.pip [hunk 4]
```diff
@@ -0,0 +1,2 @@
+udata>=3.2.0
+udata-front>=1.0.0
```
Comment (maudetes) [inline requirements/install.pip:2]: Do we need udata-front requirement?

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+setup(
+    name='udata-transport',
+    version=__import__('udata_transport').__version__,
+    description=__import__('udata_transport').__description__,
+    long_description=long_description,
+    long_description_content_type='text/markdown',
+    url='https://github.com/opendatateam/udata-transport',
+    author='Opendata Team',
+    author_email='contact@opendata.team',
```
Comment (maudetes) [inline setup.py]:
```suggestion
    author_email='opendatateam@data.gouv.fr',
```

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+from setuptools import setup, find_packages
+
+RE_REQUIREMENT = re.compile(r'^\s*-r\s*(?P<filename>.*)$')
+RE_BADGE = re.compile(r'^\[\!\[(?P<text>[^\]]+)\]\[(?P<badge>[^\]]+)\]\]\[(?P<target>[^\]]+)\]$', re.M)
+
+BADGES_TO_KEEP = []
+
+
+def md(filename):
```
Comment (maudetes) [inline setup.py]: Most of this additionnal logic had been simplified in https://github.com/opendatateam/udata/blob/4a2b564ac70f5f5bdfb5e42217f93c9be31d51e9/setup.py. Maybe we can use this simplified logic here as well.

File: tests/test_tasks.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import pytest
+import requests_mock
+
+from udata.core.dataset.factories import DatasetFactory
+
+from udata_transport.tasks import map_transport_datasets
+
+MOCK_URL = 'http://reco.net'
```
Comment (maudetes) [inline tests/test_tasks.py]: I guess it's not needed?

File: tests/test_tasks.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+
+@pytest.fixture
+def mock_response():
+    return [
+        {
+            "datagouv_id": "61fd29da29ea95c7bc0e1211",
```
Comment (maudetes) [inline tests/test_tasks.py:13]: `id` and `datagouv_id` are the same?

File: tests/test_views.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import pytest
+
+from flask import render_template_string, g
```
Comment (maudetes) [inline tests/test_views.py]:
```suggestion
from flask import render_template_string
```

File: tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+@pytest.fixture
+def datasets():
+    return DatasetFactory.create_batch(3)
+
+
+@pytest.mark.frontend
+@pytest.mark.usefixtures('clean_db')
+@pytest.mark.usefixtures('app')
+@pytest.mark.options(THEME='gouvfr')
```
Comment (maudetes) [inline tests/test_views.py:23]: It is not need to add this theme configuration I think?

File: udata_transport/settings.py [hunk 11]
```diff
@@ -0,0 +1,6 @@
+'''
+Default settings for udata-transport
+'''
+
+
+TRANSPORT_DATASETS_URL = 'https://transport.data.gouv.fr/api/datasets'
```
Comment (maudetes) [inline udata_transport/settings.py:6]: Did we check for any info on pagination?

File: udata_transport/tasks.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        return
+
+    response = requests.get(source)
+    if response.status_code != 200:
+        error('Remote platform unreachable.')
+        return
+    results_list = response.json()
+    for dataset in results_list:
+        process_dataset(dataset)
+
```
Comment (maudetes) [inline udata_transport/tasks.py]:
```suggestion
    success(f'Done. {len(results_list)} datasets mapped to transport')
```

**Review by maudetes — COMMENTED** [review]
Nice!

I have some minor feedbacks. Could we also complete the readme? :)

File: tests/test_tasks.py (hunk 8 above)
Comment (quaxsze) [reply tests/test_tasks.py:13]: seems but in doubt I use the `datagouv_id` returned by transport API

File: requirements/install.pip (hunk 4 above)
Comment (quaxsze) [reply requirements/install.pip:2]: We use `theme` from `udata-front`
https://github.com/opendatateam/udata-transport/blob/737922db21931f77a8e2be69bf324163605b609a/udata_transport/views.py#L3

File: tests/test_views.py (hunk 10 above)
Comment (quaxsze) [reply tests/test_views.py:23]: the `theme.render` seem to need it

File: udata_transport/tasks.py (hunk 2 above)
Comment (quaxsze) [reply udata_transport/tasks.py]: What do you mean? 😬

Comment (maudetes) [reply udata_transport/tasks.py]: Ah sorry, we should unset this extra for datasets that don't have a transport page anymore.

Comment (quaxsze) [reply udata_transport/tasks.py]: 👍

File: udata_transport/settings.py (hunk 11 above)
Comment (quaxsze) [reply udata_transport/settings.py:6]: Waiting for answer

Comment (quaxsze) [reply udata_transport/settings.py:6]: No pagination

File: udata_transport/tasks.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    if not target_dataset:
+        error(f"Dataset {dataset['id']} not found")
+        return
+    target_dataset.extras['transport_url'] = dataset['page_url']
+    target_dataset.save()
+
+
+@job("map-transport-datasets")
+def map_transport_datasets(self):
+    datasets = Dataset.objects().no_cache().timeout(False)
```
Comment (maudetes) [inline udata_transport/tasks.py]:
```suggestion
    datasets = Dataset.objects(extras__transport_url__exists=True).no_cache().timeout(False)
```

File: udata_transport/tasks.py [hunk 14]
```diff
@@ -16,6 +16,10 @@ def process_dataset(dataset):
 
 @job("map-transport-datasets")
 def map_transport_datasets(self):
+    datasets = Dataset.objects().no_cache().timeout(False)
+    for dataset in datasets:
+        dataset.extras.pop('transport_url', None)
```
Comment (maudetes) [inline udata_transport/tasks.py]: You need to save the modification as well

File: README.md [hunk 15]
```diff
@@ -1 +1,52 @@
-# udata-transport
\ No newline at end of file
+# udata-transport
+
+This plugin acts as a bridge between uData and transport.data.gouv.fr system.
```
Comment (maudetes) [inline README.md]:
```suggestion
This plugin acts as a bridge between [udata][] and [transport.data.gouv.fr](https://transport.data.gouv.fr/) system.
```

File: README.md [hunk 16]
```diff
@@ -1 +1,52 @@
...
+
+## Compatibility
+
+**udata-transport** requires Python 3.7+ and [uData][].
```
Comment (maudetes) [inline README.md]:
```suggestion
**udata-transport** requires Python 3.7+ and [udata][].
```

File: README.md [hunk 17]
```diff
@@ -1 +1,52 @@
...
+
+## Installation
+
+Install [uData][].
```
Comment (maudetes) [inline README.md]:
```suggestion
Install [udata][].
```

File: README.md [hunk 18]
```diff
@@ -1 +1,52 @@
...
+```
+
+Modify your local configuration file of **udata** (typically, `udata.cfg`) as following:
+
+```python
+PLUGINS = ['transport']
+TRANSPORT_DATASETS_URL = 'https://path/to/transport.data.gouv.fr/datasets/api'
+```
+
+- `TRANSPORT_DATASETS_URL`: The URL of the API endpoint listing datasets on transport.data.gouv.fr. _Default_: `None`
```
Comment (maudetes) [inline README.md]: Default is not None

File: README.md [hunk 19]
```diff
@@ -1 +1,52 @@
...
+```json
+{
+"id": "dataset-id",
+"extras": {
+    "transport:url": "https://path/to/datasets"
+  }
+}
+```
+
+[uData]: https://github.com/opendatateam/udata
```
Comment (maudetes) [inline README.md]:
```suggestion
[udata]: https://github.com/opendatateam/udata
```

**Review by maudetes — COMMENTED** [review]
Some minor feedbacks. I think we're almost good!

**Review by maudetes — APPROVED** [review]
Nice! :bike:

**Pull request merged**